internal/merkle/inmemory/reference_test.go

```diff
@@ -0,0 +1,206 @@
+// for testing correctness of other more flexible and performant algorithms,
+// such as the in-memory Tree type and compact ranges.
+//
+// [1] https://datatracker.ietf.org/doc/html/rfc6962#section-2
+
+// refRootHash returns the root hash of a Merkle tree with the given entries.
+// This is a reference implementation for cross-checking.
+func refRootHash(entries [][]byte, hasher merkle.LogHasher) []byte {
+	if len(entries) == 0 {
+		return hasher.EmptyRoot()
+	}
+	if len(entries) == 1 {
+		return hasher.HashLeaf(entries[0])
+	}
+	split := downToPowerOfTwo(uint64(len(entries)))
+	return hasher.HashChildren(
+		refRootHash(entries[:split], hasher),
+		refRootHash(entries[split:], hasher))
+}
+
+// refInclusionProof returns the inclusion proof for the given leaf index in a
+// Merkle tree with the given entries. This is a reference implementation for
+// cross-checking.
+func refInclusionProof(entries [][]byte, index uint64, hasher merkle.LogHasher) [][]byte {
+	size := uint64(len(entries))
+	if size == 1 || index >= size {
+		return nil
+	}
+	split := downToPowerOfTwo(size)
+	if index < split {
+		return append(
+			refInclusionProof(entries[:split], index, hasher),
+			refRootHash(entries[split:], hasher))
+	}
+	return append(
+		refInclusionProof(entries[split:], index-split, hasher),
+		refRootHash(entries[:split], hasher))
+}
+
+// refConsistencyProof returns the consistency proof for the two tree sizes, in
+// a Merkle tree with the given entries. This is a reference implementation for
+// cross-checking.
+func refConsistencyProof(entries [][]byte, size2, size1 uint64, hasher merkle.LogHasher, haveRoot1 bool) [][]byte {
+	if size1 == 0 || size1 > size2 {
+		return nil
+	}
+	// Consistency proof for two equal sizes is empty.
+	if size1 == size2 {
+		// Record the hash of this subtree if it's not the root for which the proof
+		// was originally requested (which happens when size1 is a power of 2).
+		if !haveRoot1 {
+			return [][]byte{refRootHash(entries[:size1], hasher)}
+		}
+		return nil
+	}
+
+	// At this point: 0 < size1 < size2.
+	split := downToPowerOfTwo(size2)
+	if size1 <= split {
+		// Root of size1 is in the left subtree of size2. Prove that the left
+		// subtrees are consistent, and record the hash of the right subtree (only
+		// present in size2).
+		return append(
+			refConsistencyProof(entries[:split], split, size1, hasher, haveRoot1),
+			refRootHash(entries[split:], hasher))
+	}
+
+	// Root of size1 is at the same level as size2 root. Prove that the right
+	// subtrees are consistent. The right subtree doesn't contain the root of
+	// size1, so set haveRoot1 = false. Record the hash of the left subtree
+	// (equal in both trees).
+	return append(
+		refConsistencyProof(entries[split:], size2-split, size1-split, hasher, false),
+		refRootHash(entries[:split], hasher))
+}
+
+// downToPowerOfTwo returns the largest power of two smaller than x.
+func downToPowerOfTwo(x uint64) uint64 {
```

Comment:
May I suggest sqrtFloor as an alternative name?

Reply:
I see where your idea comes from, but it's not quite a square root, it returns the highest power of 2 which is strictly less than x.

E.g., for 9 it returns 8, whereas the square root of 9 is 3.

This name is how this func was called before the refactoring.

Reply:
Ah, I did not look at this carefully. Fair enough.